### PR TITLE
First version of the data objects blog post

### PR DESCRIPTION
**Pull request by pcreux**
- [x] Write
- [x] Wording Review - Thank you @patdryburgh @mstopham
- [ ] Technical Review :smile_cat: 
- [x] social description
- [x] social image

screencapture-localhost-4000-2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects-html-1438384165109

### CODE REVIEW

**Comment by patdryburgh** [comment]
## Suggested Edits

*Paragraph 1:* 
…and a json payload is a hash with nested arrays and hashes

*Paragraph 2:*
There are times where those data structures become too complex and we have to dig deeper into the codebase…

*Paragraph 3:*
Data objects contain data; they don't implement any behaviour.

*Paragraph 5:*
Documentation such as this is a great gift for other developers (and your future-self).

*Paragraph 7:*
I fell in love with Virtus a couple years ago.

*Paragraph 12:*
…returns information about an email you sent including sender, subject, opens, and clicks as well as a list of all open and click events.

*Paragraph 13:*
…and look up the online documentation to determine what's available.

*Paragraph 17:*
“So the next time you deal with a complex data structure, do yourself a favour and turn it into a Data Object with Virtus.

**Comment by pcreux** [comment]
@patdryburgh Thank you!

**Comment by pcreux** [comment]
A few more comments from @mstopham 

```
1) goto = go-to
2) developer happiness = developer-happiness
3) ...recursively flattened wait what?! = recursively flattened...wait, what?!
4) virtus, the data object best friend = virtus, the data object('s) best friend<--?
5) oh surprise = oh, surprise
6) data object = Data Object (think you have in one of the headers lower case and then throughout the post you have upper case
7) reason about = reason with <--?
```

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  - ruby
+  - virtus
+shared_square_image:
+shared_description:
+draft: false
+published: true
+---
+
+
+In Ruby, hashes and arrays are the go-to data structures. For example, Rails turns http params into a hash and a json payload is a hash with nested arrays and hashes. The flexibility those primitives offer is undeniable and is key to developer happiness.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: `The flexibility those primitives offer is undeniable`
--> The flexibility offered by those primitives is undeniable

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+There are times where those data structures become too complex and we have to dig deeper into the codebase to figure what a hash is supposed to contain. This is where Data Objects can help us a great deal.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: > figure what a hash is supposed to contain
--> figure out

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+Data Objects contain data; they don’t implement any behaviour. A Struct is the simplest Data Object you can think of. It has a defined set of attributes that are publicly available as instance methods.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: > they don't implement any behaviour.

Could remove "any". "they don't implement behaviour" sounds better IMO.

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+{% highlight ruby %}
+class Email < Struct.new(:from, :to, :subject, :body)
+end
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: I think it's actually better to make a variable declaration in this case. The `Email` will have a small list of ancestors.

```ruby
irb(main):003:0> class A < Struct.new(:a)
irb(main):004:1> end
=> nil
irb(main):005:0> A.ancestors
=> [A, #<Class:0x007f83de1d5f10>, Struct, Enumerable, Object, Kernel, BasicObject]
irb(main):006:0> A = Struct.new(:a)
=> A
irb(main):007:0> A.ancestors
=> [A, Struct, Enumerable, Object, Kernel, BasicObject]
```

:neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard: :neckbeard:

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+{% endhighlight %}
+
+The main benefit of Data Objects is that they explicitly define the attributes available.
+
+Documentation such as this is a great gift for other developers (and your future-self). A quick look at the Data Object definition tells us what attributes are available. No need to go through those four classes where a hash that's returned by a third-party API is transformed, filtered, reduced, deleted, symbolized keys and recursively flattened... wait, what?!
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: > Documentation such as this is a great gift for other developers (and your future-self).

"This kind of self-documentation is a great gift for..."

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+{% highlight ruby %}
+User.new(email: "bob@example.com", created_at: "2015-01-01 12:00:00")
+
+# => #<User:0x007f97e0cca928
+#       @email="bob@example.com",
+#       @created_at=2015-01-01 12:00:00 -0800,
+#       @verified=false>
+{% endhighlight %}
+
+The syntax to define attributes works as an excellent documentation. Ruby is not a typed language, but Virtus provides *some sort of a feel of optional typing*.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: > The syntax to define attributes works as an excellent documentation.

"... works as excellent documentation."

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+As you can see in the example above, Virtus attempts to coerce values to the type passed in. This is great for consuming APIs or http params: `Strings` will be converted to `Integer`, `Boolean` or `Time` if you’ve defined your attribute to be so. Even better, nested arrays and hashes can be coerced to other Virtus model.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: > to other Virtus model.

"... to another Virtus model."

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+{% highlight ruby %}
+user.attributes
+# => {:email=>"bob@example.com", :verified=>false, :created_at=>2015-01-01 12:00:00 -0800}
+{% endhighlight %}
+
+## Wrapping API responses with Virtus
+
+Let’s take the [Mandrill API](https://mandrillapp.com/api/docs/index.ruby.html) as an example here. The
+[end-point `/messages/info.json`](https://mandrillapp.com/api/docs/messages.JSON.html#method=info) returns information about an email you sent including sender, subject, opens, clicks as well as all open and click events.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: > including sender, subject, opens, clicks as well as all open and click events.

"... opens, clicks and corresponding events".

#gabesuggests2015

#gabesuggests2015

File: _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+The ruby wrapper turns that json into (oh, surprise!) a large hash. You could query the hash via `reponse.fetch('metadata').fetch('user_id')` and look up the online documentation to determine what's available. Let's create a Data Object to wrap the API responses here. Using Virtus and some code-editing-fu it takes a couple of seconds to turn the documentation into a Data Object class.
```
Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: Should we capitalize json to JSON? I dunno.

Comment (garbles) [inline _posts/2015-07-31-be-nice-to-others-and-your-future-self-use-data-objects.md]: `code-editing-fu` :clap:

**Comment by pcreux** [comment]
@garbles Thank you! :boom:

**Pull request merged**